Classes/Image Upload/ImageUploadTableViewController.swift

```diff
@@ -0,0 +1,193 @@
+    private func setRightBarItemSpinning() {
+        let activity = UIActivityIndicatorView(activityIndicatorStyle: .gray)
+        activity.startAnimating()
+        navigationItem.rightBarButtonItem = UIBarButtonItem(customView: activity)
+    }
+    
+    /// Sets the right bar button item to have a checkmark, enabling the user to upload the image
+    private func setRightBarItemIdle() {
+        let item = UIBarButtonItem(
+            image: UIImage(named: "check"),
+            style: .plain,
+            target: self,
+            action: #selector(didPressTick)
+        )
+        
+        item.tintColor = Styles.Colors.Green.medium.color
+        item.accessibilityLabel = NSLocalizedString("Upload", comment: "")
+        navigationItem.rightBarButtonItem = item
+    }
+    
+    @IBAction func didPressClose() {
+        let dismissBlock = {
+            self.dismiss(animated: true)
+        }
+        
+        if titleText == nil && descriptionText == nil {
+            dismissBlock()
+            return
+        }
+        
+        let title = NSLocalizedString("Unsaved Changes", comment: "Image Upload - Dismiss w/ Unsaved Changes Title")
+        let message = NSLocalizedString("Are you sure you don't want to upload this image? Your title & description will not be saved!", comment: "Image Upload - Cancel w/ Dismiss Changes Message")
+        
+        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
+        alert.addActions([
+            AlertAction.goBack(),
+            AlertAction.discard { _ in
+                dismissBlock()
+            }
+            ])
+        
+        present(alert, animated: true, completion: nil)
+    }
+    
+    @IBAction func didPressTick() {
+        // Set the right bar item back to the spinner so they can't spam the button, and to indicate something is happening
+        setRightBarItemSpinning()
+        
+        // Should never caught in here as the button will be disabled in this situation
+        guard let compressionData = compressionData else {
+            ToastManager.showGenericError()
+            return
+        }
+        
+        // Check that we have enough "tokens" to actually upload the image
+        client.canUploadImage { [weak self] success in
+            // Ensure that we do have enough tokens, otherwise remove the upload button
+            guard success else {
+                ToastManager.showError(message: NSLocalizedString("Rate Limit reached, cannot upload!", comment: ""))
+                self?.navigationItem.rightBarButtonItem = nil
+                return
+            }
+            
+            var name = "GitHawk Upload"
+            
+            if let username = self?.username {
+                name += " by \(username)"
+            }
+            
+            // Ensure the upload step is on the background thread
+            DispatchQueue.global(qos: .userInitiated).async {
```

Comment:
Shouldn't the `request(...)` already happen in the background? This isn't necessary, right? At the very least, we should hide the async inside `uploadImage(...)`

Reply:
Probably leftover from when I had the compression task just before the network request. Will neaten.